### PR TITLE
Create the evaluation machine

### PR DESCRIPTION
**Pull request by DarkDrek**
Implements #55 

Noteworthy changes from `cargo bench`

arithmetic              time:   [22.523 us 22.605 us 22.704 us]
                        change: [-5.8726% -5.4123% -4.9390%] (p = 0.00 < 0.05)
                        Performance has improved.

gcd                     time:   [1.1096 ms 1.1127 ms 1.1159 ms]
                        change: [+41.983% +44.259% +46.070%] (p = 0.00 < 0.05)
                        Performance has regressed.

fancy_max               time:   [3.5319 us 3.5494 us 3.5798 us]
                        change: [+27.366% +30.815% +33.250%] (p = 0.00 < 0.05)
                        Performance has regressed.

step                    time:   [3.1044 us 3.1353 us 3.1622 us]
                        change: [+73.727% +81.375% +87.492%] (p = 0.00 < 0.05)
                        Performance has regressed.

PS: All other were within noise threshold.

### CODE REVIEW

File: src/machine/eval.rs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        // If t1 is not an abstraction, evaluate it.
+        } else {
+            (self.step_in_place(t1.as_mut()), Term::Fix(t1))
+        }
+    }
+}
+
+/// Evaluation step for beta reduction ((λ. body) arg)
+#[inline(always)]
+fn step_beta_reduction(mut body: Box<Term>, mut arg: Box<Term>) -> (bool, Term) {
```
Comment (pvdrz) [inline src/machine/eval.rs]: I think you can change this one to be a method of `Machine` too for the sake of consistency.

**Review by pvdrz — COMMENTED** [review]
Hmm I wonder why the regressions. `step` and `fancy_max` don't worry me because they are super fast, so an 81% increase for a 3 microseconds time is really not a big deal. However `gcd`and `arithmetic` suprise me. I'll pull your branch and try to dig in a little.

Even then these changes are needed so they're going to happen anyway.

Update: I forgot to say something. The changes surprise me because `Machine` is a zero-sized type. So it cannot exist during runtime and Rust should optimize the code to be a bunch of function calls. Also its funny that the other heavy benchmarks weren't affected at all.

**Comment by DarkDrek** [comment]
I'm not sure if `#[inline(always)]` actually improves the performance in at least some of the methods since they are so big.

**Comment by pvdrz** [comment]
> I'm not sure if `#[inline(always)]` actually improves the performance in at least some of the methods since they are so big.

I have no idea tbh. I was thinking that actually inlining them would make the binary larger but faster to execute. Let me test it to see what happens.

**Comment by DarkDrek** [comment]
I would guess `step_in_place` may be the best function for `#[inline(always)]` but we should better have good reason to use that attribute since it is usally "a bad idea".

> Note: The rustc compiler automatically inlines functions based on internal heuristics. Incorrectly inlining functions can make the program slower, so this attribute should be used with care.

**Comment by pvdrz** [comment]
Ha, great to know

**Comment by pvdrz** [comment]
Here are the results of the benchmarks after removing inlining in master
```
arithmetic              time:   [21.417 us 21.455 us 21.506 us]                        
                        change: [-1.6868% -1.2790% -0.8335%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 16 outliers among 100 measurements (16.00%)
  4 (4.00%) high mild
  12 (12.00%) high severe

logic                   time:   [43.711 us 43.802 us 43.911 us]                   
                        change: [-0.9175% -0.2730% +0.3576%] (p = 0.42 > 0.05)
                        No change in performance detected.
Found 14 outliers among 100 measurements (14.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  10 (10.00%) high severe

factorial               time:   [57.040 us 57.172 us 57.340 us]                      
                        change: [-4.6175% -3.9482% -3.3241%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  1 (1.00%) low mild
  3 (3.00%) high mild
  9 (9.00%) high severe

factorial_tail          time:   [102.82 us 103.33 us 103.88 us]                           
                        change: [-0.1605% +0.4737% +1.1440%] (p = 0.17 > 0.05)
                        No change in performance detected.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe

fibonacci               time:   [458.95 us 460.63 us 462.82 us]                      
                        change: [-1.1431% -0.5021% +0.2107%] (p = 0.16 > 0.05)
                        No change in performance detected.
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) high mild
  4 (4.00%) high severe

fibonacci_tail          time:   [88.196 us 88.319 us 88.463 us]                           
                        change: [-4.6219% -3.7119% -2.9308%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  4 (4.00%) high mild
  7 (7.00%) high severe

gcd                     time:   [740.96 us 742.56 us 744.42 us]                
                        change: [-0.8543% -0.2823% +0.2979%] (p = 0.36 > 0.05)
                        No change in performance detected.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

ackermann               time:   [362.35 us 363.95 us 365.88 us]                      
                        change: [-1.8157% -0.5344% +0.5222%] (p = 0.42 > 0.05)
                        No change in performance detected.
Found 19 outliers among 100 measurements (19.00%)
  1 (1.00%) low mild
  6 (6.00%) high mild
  12 (12.00%) high severe

calling                 time:   [140.93 us 141.22 us 141.54 us]                    
                        change: [+2.5877% +3.0769% +3.5324%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  3 (3.00%) low mild
  4 (4.00%) high mild
  3 (3.00%) high severe

complex_calling         time:   [149.99 us 150.80 us 151.79 us]                            
                        change: [-0.8971% -0.1240% +0.6375%] (p = 0.76 > 0.05)
                        No change in performance detected.
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) low mild
  5 (5.00%) high mild
  5 (5.00%) high severe

fancy_max               time:   [2.2160 us 2.2225 us 2.2302 us]                       
                        change: [+0.6212% +1.5434% +2.4600%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 12 outliers among 100 measurements (12.00%)
  6 (6.00%) high mild
  6 (6.00%) high severe

step                    time:   [1.3253 us 1.3272 us 1.3293 us]                  
                        change: [+4.7587% +5.2840% +5.8316%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 7 outliers among 100 measurements (7.00%)
  4 (4.00%) high mild
  3 (3.00%) high severe
```
I'd say there are more `-` than `+` everywhere. So you might be in the right track.

**Comment by DarkDrek** [comment]
I also tried it in my branch locally and got 

```
factorial_tail          time:   [119.40 us 119.88 us 120.57 us]
                        change: [-15.757% -15.313% -14.848%] (p = 0.00 < 0.05)
gcd                     time:   [758.30 us 762.77 us 768.87 us]
                        change: [-31.928% -31.543% -31.060%] (p = 0.00 < 0.05)
fancy_max               time:   [5.1950 us 5.4056 us 5.5769 us]
                        change: [+28.236% +34.159% +40.300%] (p = 0.00 < 0.05)
step                    time:   [2.0688 us 2.0772 us 2.0888 us]
                        change: [-29.761% -27.110% -24.241%] (p = 0.00 < 0.05)
```

**Review by pvdrz — APPROVED** [review]

**Comment by pvdrz** [comment]
I'd say this is a good trade-off. There might be other optimizations that I want to test but we can do it in other PR. Thanks for this!

**Pull request merged**